Arithmetic.md

```diff
@@ -0,0 +1,120 @@
+# Arithmetic in hacspec
+
+The hacpsec library offers a set of tools to perform arithmetic operations on different objects.
+All objects can be either secret or public.
+
+Any arithmetic object must implement the following trait.
+Objects might leave functions unimplemented.
+```Rust
+pub trait Numeric:
+    Add<Self, Output = Self>
+    + AddAssign
+    + Sub<Self, Output = Self>
+    + SubAssign
+    + Mul<Self, Output = Self>
+    + MulAssign
+    + BitXor<Self, Output = Self>
+    + BitXorAssign
+    + BitOr<Self, Output = Self>
+    + BitOrAssign
+    + BitAnd<Self, Output = Self>
+    + BitAndAssign
+    + Shl<u32, Output = Self>
+    + ShlAssign
+    + Shr<u32, Output = Self>
+    + ShrAssign
+    + Default
+    + Copy
+    + Debug
+{
+    /// Return largest value that can be represented.
+    fn max() -> Self;
+    
+    /// (self - rhs) % n.
+    fn sub_mod(self, rhs: Self, n: Self) -> Self;
+    /// `(self + rhs) % n`
+    fn add_mod(self, rhs: Self, n: Self) -> Self;
+    /// `(self * rhs) % n`
+    fn mul_mod(self, rhs: Self, n: Self) -> Self;
+    /// Division.
+    fn div(self, rhs: Self) -> Self;
+    /// `self % n`
+    fn rem(self, n: Self) -> Self;
+    /// Invert self modulo n.
+    fn inv(self, n: Self) -> Self;
+    /// `|self|`
+    fn abs(self) -> Self;
+
+    // Comparison functions returning bool.
+    fn equal(self, other: Self) -> bool;
+    fn greater_than(self, other: Self) -> bool;
+    fn greater_than_or_qual(self, other: Self) -> bool;
+    fn less_than(self, other: Self) -> bool;
+    fn less_than_or_equal(self, other: Self) -> bool;
+
+    // Comparison functions returning a bit mask (0x0..0 or 0xF..F).
+    fn equal_bm(self, other: Self) -> Self;
+    fn greater_than_bm(self, other: Self) -> Self;
+    fn greater_than_or_qual_bm(self, other: Self) -> Self;
+    fn less_than_bm(self, other: Self) -> Self;
+    fn less_than_or_equal_bm(self, other: Self) -> Self;
+}
+```
+
+Further, every arithmetic object must implement direct instantiation options through `From` taking `u128` or `&[u128]`.
```

Comment:
Thanks for the writeup Franziskus ! So why did you choose `u128` for these direct instantiations ? Is it to allow instantiation with integer literals in the code like `0` or `1` ? Furthermore, I think we should use `Seq<u128>` instead of `&[u128]` as we don't want to deal with primitive Rust slices in Hacspec.

Reply:
> Is it to allow instantiation with integer literals in the code like 0 or 1 ? 

Yep, I think we should whenever possible allow instantiations with literals.

> Furthermore, I think we should use Seq<u128> instead of &[u128] as we don't want to deal with primitive Rust slices in Hacspec.

I'm not sure what's the best way is here. `Seq<u128>` would be the hacspec way, but also pretty cumbersome. And we said we allow slices for `Seq` instantiation already. So I figured this is a similar case. But I think I could live with `Seq` if we can do `Seq<u128>::from(&[u128])` or for example `Seq<(usize, u128)>::from(&[(usize, u128)])` for sparse polynomials.

Reply:
Hum I see your point, I guess we can syntactically restrict (clippy) slices from only living inside `from` function calls. Thanks !

Reply:
Thanks for that as well @franziskuskiefer ! Looks good to me =) Should we add also rotate_left/rotate_right as well ?
As @denismerigoux said, that shoud do it via clippy. We should track what we'll allow in "constructors"/conversion functions/macro arguments but forbid in the rest of the language.

Reply:
> Should we add also rotate_left/rotate_right as well ?

Good point! I also noticed and already added it locally :slightly_smiling_face: